authentication/managing_cloud_provider_credentials/cco-mode-sts.adoc

```diff
@@ -0,0 +1,734 @@
+
+*** `<public-signing-key-exponent>` is generated from the public key with:
++
+[source,terminal]
+----
+$ printf "%016x" $(openssl rsa -pubin -in sa-signer.pub -noout -text | grep Exponent | awk '{ print $2 }') |  awk '{ sub(/(00)+/, "", $1); print $1 }' | xxd -r -p | base64 -w0 | tr '/+' '_-' | tr -d '='
+----
++
+This command extracts decimal representation of public key exponent, prints it as hex with a padded `0` if needed, removes leading `00` pairs, converts the ASCII hex to binary, encodes the data with base64 encoding, and then changes base64-encoded output to use only characters that can be used in a URL.
+
+.. Save a file named `openid-configuration` that contains the following information:
++
+[source,json]
+----
+{
+	"issuer": "$OPENID_BUCKET_URL",
+	"jwks_uri": "${OPENID_BUCKET_URL}/keys.json",
+    "response_types_supported": [
+        "id_token"
+    ],
+    "subject_types_supported": [
+        "public"
+    ],
+    "id_token_signing_alg_values_supported": [
+        "RS256"
+    ],
+    "claims_supported": [
+        "aud",
+        "exp",
+        "sub",
+        "iat",
+        "iss",
+        "sub"
+    ]
+}
+----
+
+. Upload the OIDC configuration:
++
+[source,terminal]
+----
+$ aws s3api put-object --bucket <oidc-bucket-name> --key keys.json --body ./keys.json
+----
++
+[source,terminal]
+----
+$ aws s3api put-object --bucket <oidc-bucket-name> --key '.well-known/openid-configuration' --body ./openid-configuration
+----
+
+. Allow the AWS IAM OIDC identity provider to read these files:
++
+[source,terminal]
+----
+$ aws s3api put-object-acl --bucket <oidc-bucket-name> --key keys.json --acl public-read
+----
++
+[source,terminal]
+----
+$ aws s3api put-object-acl --bucket <oidc-bucket-name> --key '.well-known/openid-configuration' --acl public-read
+----
+
+. Create an AWS IAM OIDC identity provider:
+
+.. Get the certificate chain from the server that hosts the OIDC configuration:
++
+[source,terminal]
+----
+$ echo | openssl s_client -servername s3.$<aws-region>.amazonaws.com -connect s3.$<aws-region>.amazonaws.com:443 -showcerts 2>/dev/null | awk '/BEGIN/,/END/{ if(/BEGIN/){a++}; out="cert"a".pem"; print >out}'
+----
+
+.. Calculate the fingerprint for the certificate at the root of the chain:
++
+[source,terminal]
+----
+$ export BUCKET_FINGERPRINT=$(openssl x509 -in cert<number>.pem -fingerprint -noout | sed -e 's/.*Fingerprint=//' -e 's/://g')
+----
++
+Where `<number>` is the highest number in the files that were saved. For example, if two is the highest number in the files that were saved, use `cert2.pem`.
+
+.. Create the identity provider:
++
+[source,terminal]
+----
+$ aws iam create-open-id-connect-provider --url $OPENID_BUCKET_URL --thumbprint-list $BUCKET_FINGERPRINT
+----
+
+.. Save the returned ARN of the newly created identity provider to `$AWS_IAM_OPENID_ARN`.
```

Comment:
My ignorance: Not sure how to do this step.

Reply:
I think this is another env variable? @joelddiaz

Reply:
Again, implied steps. Just telling someone to keep track of this details as it will be needed in a future step.

Reply:
To attempt to make this clearer but not lose people I have:
> Retain the returned ARN of the newly created identity provider. This ARN is later referred to as `<aws-iam-openid-arn>`.

And then where is it used:
```
[...]
{
                    "Effect": "Allow",
                    "Principal": {
                        "Federated": "<aws-iam-openid-arn>"
                    },
[...]
```
> Where `<aws-iam-openid-arn>` is the returned ARN of the newly created identity provider.

Hopefully that keeps anyone from getting lost!